gsplat/cuda/csrc/utils.cuh

```diff
@@ -168,6 +171,60 @@ inline __device__ void persp_proj(
     );
     cov2d = J * cov3d * glm::transpose(J);
     mean2d = vec2<T>({fx * x * rz + cx, fy * y * rz + cy});
+
+    // calculate the ray space intersection plane.
+    auto length = [](T x, T y, T z) { return sqrt(x*x+y*y+z*z); };
+    mat3<T> cov3d_eigen_vector;
+	vec3<T> cov3d_eigen_value;
+	int D = glm_modification::findEigenvaluesSymReal(cov3d,cov3d_eigen_value,cov3d_eigen_vector);
+    unsigned int min_id = cov3d_eigen_value[0]>cov3d_eigen_value[1]? (cov3d_eigen_value[1]>cov3d_eigen_value[2]?2:1):(cov3d_eigen_value[0]>cov3d_eigen_value[2]?2:0);
+    mat3<T> cov3d_inv;
+	bool well_conditioned = cov3d_eigen_value[min_id]>1E-8;
+	vec3<T> eigenvector_min;
+	if(well_conditioned)
+	{
+		mat3<T> diag = mat3<T>( 1/cov3d_eigen_value[0], 0, 0,
+                                0, 1/cov3d_eigen_value[1], 0,
+                                0, 0, 1/cov3d_eigen_value[2] );
+		cov3d_inv = cov3d_eigen_vector * diag * glm::transpose(cov3d_eigen_vector);
+	}
+	else
+	{
+		eigenvector_min = cov3d_eigen_vector[min_id];
+		cov3d_inv = glm::outerProduct(eigenvector_min,eigenvector_min);
+	}
+    vec3<T> uvh = {u, v, 1};
+    vec3<T> Cinv_uvh = cov3d_inv * uvh;
+    if(length(Cinv_uvh.x, Cinv_uvh.y, Cinv_uvh.z) < 1E-12 || D ==0)
+    {
+        normal = {0, 0, 0};
+        ray_plane = {0, 0};
+    }
+    else
+    {
+        T l = length(tx, ty, z);
+        mat3<T> nJ_T = glm::mat3(rz, 0.f, -tx * rz2,             // 1st column
+                                0.f, rz, -ty * rz2,              // 2nd column
+                                tx/l, ty/l, z/l                  // 3rd column
+        );
+        T uu = u * u;
+        T vv = v * v;
+        T uv = u * v;
+
+        mat3x2<T> nJ_inv_T = mat3x2<T>(vv + 1, -uv,          // 1st column
+                                       -uv, uu + 1,          // 2nd column
+                                        -u, -v               // 3nd column
+        );
+        T factor = l / (uu + vv + 1);
+        vec3<T> Cinv_uvh_n = glm::normalize(Cinv_uvh);
+        T u_Cinv_u_n_clmap = max(glm::dot(Cinv_uvh_n, uvh), 1E-7);
```

Comment:
Hello @BaowenZ , I am a little bit lost here and can not align the code with the equation in original paper. Is there any hint?

Reply:
This is related to further derivation of the rendering equation. The plane vector is $\frac{\Sigma'^{-1}v'}{v'^T\Sigma'^{-1}v'}$, where $\Sigma'={J}{\Sigma}^\top{J}^\top$ and $v'=(0,0,1)^\top$. After pluging $\Sigma'$ and $v'$, the $J^{-1}v'$ is the last column of $J^{-1}$. After substracting $z$ from each term to make x, y become u, v, and reduction, we can get the code.